### PR TITLE
feat: add bluetooth backend abstraction & support devices with multiple measurements

### LINKED ISSUE #352: App always imports the first measurement instead the newest
### Description

With Medisana BU550 connect the app imports not the newest measurement. Always the oldest one is saved.

Btw: Thank you for development!

### Android version

Android 11

### Reproduction steps

```bash
1. Delete memory of the BU550 connect
2. Make the first measurement and import it.
3. Make another measurement and try to import.
4. You will see that the first measurement was imported.
```

### Screenshots & Settings backup

_No response_

### App debug info

_No response_

### LINKED ISSUE #423: Use different bluetooth library that supports all major platforms
### What you want:

Use a bluetooth library that supports all major platforms, not just Android/iOS.

Possible alternatives:
- https://pub.dev/packages/universal_ble
- https://pub.dev/packages/bluetooth_low_energy

I can help implement this if you'd like, but like your approval first :) 

Also for this and the other issues, maybe we can create a github project with tickets you'd like help with? Or at least add a label 'help-wanted' to those tickets and finalize the feature description. Not sure how much time you have and/or want to spend on this app but I can help with some tickets if they have a clear ask and you agree with them :)

### Why you want it:

The maintainer of https://github.com/chipweinberger/flutter_blue_plus has rejected PR's to add Linux & Web support to FBP, in the issue about adding Linux Support he says support will happen via wrappers but that implementation has big disadvantages.

### PR DESCRIPTION
**Pull request by pimlie**
Resolves #423
Resolves #352

> This is a draft pr for now, need to:
> - [x] review it a bit more (in general and Linux compatibility specifically)
> - [x] check my code comments and
> - [x] at minimum add some tests for measurement_multiple.dart
>

- To resolve 423 I decided to add an abstraction layer for the bluetooth implementation. I wasn't fully sure whether the `bluetooth_low_energy` package will be the best (the maintainer is quite responsive tho) so with this abstraction we should be able to quickly switch between backends if needed.
- I've added a BleReadMultiple read state that will shows a ListView with all returned measurements by the device
  - ⚠️ I've noticed that one of my devices has two reading modes. You either 1) take a measurement after which it enables bluetooth and you receive the one measurement you just did, or 2) you can hold the BT button on the device for 3s after which it turns into download mode. But reading in download mode seems to remove all data from the device memory, this feature seems to exists to sync all data with the app from the manufacturer which would store all your measurements in the cloud. My current implementation will only read one measurements, so that would mean users could loose other measurements. I'm happy to change the current implementation to sync/import all but lets create a ticket that defines the desired behaviour then first and create a new PR for that behaviour.
  - The ListView has no limit at the moment, even though some devices could store up to 100 measurements per user. In those cases then all those 100 measurements would be shown atm, again happy to change this behaviour but let's create a ticket to define the desired behaviour first and create a new PR for that behaviour
  - I did not check what happens when a device supports multiple users, and whether it always downloads all measurements from all users (which I think it does). If so then the 100 measurements above could also be 2 * 100...
- I've added logging support through the official `logging/logging` package. This provide much more granular log levels and would also make it easier to support in-app logging etc if we ever decide we want that (e.g. think debug support in beta builds). (I added this because I wanted to be able to see the bluetooth_low_energy logs which also used logging)
- I implemented the bluetooth abstraction layer with the goal that implementing the happy path of reading bluetooth data is as straight forward as possible. By that I mean I preferred using Future's instead of Stream's, using StreamTransformers if possible and I only implemented what we currently need
- The FlutterBluePlus backend is still using your FlutterBluePlusMockable class, this means that we can re-use the existing tests and also test most of the new backend logic, even if we use the bluetooth_low_energy package when building the app.
- Added [BLUETOOTH.md](https://github.com/pimlie/blood-pressure-monitor-fl/blob/feat-use-ble-package/BLUETOOTH.md) so we can maintain a list of supported devices and their features. In a next pr we can expand the list to f.e. also list time support etc

### CODE REVIEW

**Comment by NobodyForNothing** [comment]
That looks great so far! You kept that code really tidy. I will be happy to properly review this once it's ready.

I noticed your code style is fairly simmilar to what you see in java. Specifically: In dart comments work a bit differently (e.g references to other classes in square brackets, the first line in the comment is a short summary, details come after a blank line). There is a [great doc page](https://dart.dev/effective-dart/documentation) about it with notes on how to  write helpful comments that can act as a general guideline. I just wanted to give you an heads up about it, as this might come unexpected for someone new to dart.

File: app/lib/features/bluetooth/backend/bluetooth_low_energy/ble_device.dart
Comment (NobodyForNothing) [reply app/lib/features/bluetooth/backend/bluetooth_low_energy/ble_device.dart:49]: I personally dislike too many abstractions. They often make the code less efficient and can lead to bugs when someone reuses the wrong bit of code.

This is a matter of preference, but I like to read my code top-to-bottom unless that would be really inefficient or break the controll-flow.

File: app/lib/features/bluetooth/backend/bluetooth_low_energy/ble_device.dart
Comment (NobodyForNothing) [inline app/lib/features/bluetooth/backend/bluetooth_low_energy/ble_device.dart:51]: When this hints at an developer error asserts make sure it is caught.
```suggestion
      assert(false, 'Device not connect.Call device.connect() first.');
```

File: app/lib/features/bluetooth/backend/bluetooth_low_energy/ble_device.dart
Comment (NobodyForNothing) [reply app/lib/features/bluetooth/backend/bluetooth_low_energy/ble_device.dart:69]: notifications and indications are very similar, I read somewhere that they are often abstracted away.
But when I implemented ble, using the wrong one resulted in not reading the data successfully. So it would be best if you could test that on your device.

File: app/lib/features/bluetooth/ui/measurement_failure.dart
Comment (NobodyForNothing) [reply app/lib/features/bluetooth/ui/measurement_failure.dart:18]: I think that's fine.

File: app/lib/features/input/add_measurement_dialoge.dart
Comment (NobodyForNothing) [reply app/lib/features/input/add_measurement_dialoge.dart:257]: I think I suggested a pattern in the bluetooth_backend file.

File: app/lib/logging.dart
Comment (NobodyForNothing) [reply app/lib/logging.dart:66]: Grepping around stack traces isn't very useful outside debug builds since class names will be obfuscated in releases. Stack traces are useful as they are imo.

**Review by NobodyForNothing — CHANGES_REQUESTED** [review]
Thanks allot for this PR. I added some thoughts to the first few files and tried to answer your questions. It would be great if you could consider them and apply these lessons to other files.

Don't feel discouraged by the amount or density, I tried to point everything out. You found some really great solutions, that I didn't comment on.

In the meantime I will try to implement apk building in PRs so you can better test this on real devices.

File: app/lib/features/bluetooth/logic/device_scan_cubit.dart
Comment (NobodyForNothing) [reply app/lib/features/bluetooth/logic/device_scan_cubit.dart:110]: Missed that one ;). I can't think of a realistic use case where you have to differentiate between two of the same device. Using an device id here wouldn't help with that since all information the user has is the name. Additionally changing this behavior would either require users to re-add their devices after an update or add complexity in the form of duplicate code to maintain.

**Comment by NobodyForNothing** [comment]
As a general note: You can drop the support for the flutter_blue_plus backend. This should drastically simplify the code.

File: app/lib/features/bluetooth/logic/device_scan_cubit.dart
Comment (NobodyForNothing) [reply app/lib/features/bluetooth/logic/device_scan_cubit.dart:110]: The user can't add arbitrary device names, but they are only shown (and most likely only know) the name. Users _could_ ensure only their device is enabled when initially adding it, but this isn't really an intuitive UX I would want to endorse.

I wouldn't want to spend too much thought on it. This edge case would only occur when multiple devices of the exact same type take measurements at the same time in close proximity. I can't think of a place where this would be the case.

If you wanted to be extra sure you could auto-select the device with the strongest signal (e.g. closest to the phone), but I don't think that is really neccessary

File: app/lib/config.dart
Comment (NobodyForNothing) [reply app/lib/config.dart:2]: Thanks for the explanation. I still don't think it's an optimal solution, but it's definitely more manageable than keeping lots of developer docs up to date.

File: app/lib/config.dart
Comment (NobodyForNothing) [reply app/lib/config.dart:4]: Could you please just move it to the logger then and add your explanation there?

File: app/lib/config.dart
Comment (NobodyForNothing) [inline app/lib/config.dart:17]: No need to track it in code when there is an open issue.
```suggestion
```

File: app/lib/features/bluetooth/backend/bluetooth_device.dart
Comment (NobodyForNothing) [reply app/lib/features/bluetooth/backend/bluetooth_device.dart:71]: Since that code is trivial in-lining is fine. Someone reading that code has an easier time understanding what the if statement does than trying to decode the function name and read the doc or jump to its definition.

File: app/lib/features/bluetooth/backend/bluetooth_device.dart
Comment (NobodyForNothing) [reply app/lib/features/bluetooth/backend/bluetooth_device.dart:74]: > Not sure I understand what you mean?

I misunderstood your code. It wasn't clear that `backendConnect` started producing events. Having backendConnect return the stream might be simpler to read.

File: app/lib/features/bluetooth/backend/bluetooth_backend.dart
Comment (NobodyForNothing) [reply app/lib/features/bluetooth/backend/bluetooth_backend.dart:21]: A factory method would be more explicit and the intuitive place to look for an implementation.

File: app/lib/features/bluetooth/backend/bluetooth_discovery.dart
Comment (NobodyForNothing) [reply app/lib/features/bluetooth/backend/bluetooth_discovery.dart:25]: Unfortunately static analysis can't catch everything so I don't want to rely on them. I use them as hints while coding and try won't introduce new warnings, but there is still some code from before using them, which is why I won't make them blocking.

**Comment by NobodyForNothing** [comment]
PRs labeled with `build-apk` are now build automatically. Please tell me whether you can work with the build result and whether you can add that label yourself.

File: app/lib/features/bluetooth/backend/bluetooth_backend.dart
Comment (NobodyForNothing) [reply app/lib/features/bluetooth/backend/bluetooth_backend.dart:16]: You're right, it doesn't work with a factory method, but you can use a static method which has similar usability.
```dart
static BluetoothManager create([bool legacyMode = false]) {
  if (isTestingEnvironment) {
    return MockBluetoothManager();
  } else if (legacyMode) {
    return FlutterBluePlusManager();
  }
  return BluetoothLowEnergyManager();
}
```

Since I don't want to annoy/pressure you during review and both take similar time: Do you want to consider and implement my suggestions yourself or should I commit formatting/pattern things myself?

File: app/lib/features/bluetooth/backend/bluetooth_connection.dart
Comment (NobodyForNothing) [reply app/lib/features/bluetooth/backend/bluetooth_connection.dart:22]: > btw, moved everything to map's

No you didn't 🙃

> afaik transform is 'more' correct so should be fine to keep it

It really makes reading the code more difficult. 

The old tests could be deleted and we can rely on implicit testing, if that's what you're concerned about. This shouldn't slow down debugging too much and might even make it faster because there are less jumps when reading the code.

**Comment by NobodyForNothing** [comment]
@pimlie Since this is kinda important for the review here again: Do you want to consider and implement my suggestions yourself or should I just commit my formatting/pattern suggestions?

File: app/lib/features/bluetooth/backend/bluetooth_service.dart
Comment (NobodyForNothing) [inline app/lib/features/bluetooth/backend/bluetooth_service.dart:14]:
```suggestion
  /// Construct abstract BluetoothUuid representation.
  BluetoothUuid({ required this.uuid }): assert(uuid.toString().length == 36) {
    assert(isBluetoothUuid);
  }
```

**Review by NobodyForNothing — CHANGES_REQUESTED** [review]
Thanks allot for your work! I commented some questions and suggestions, and pushed a few changes that shouldn't affect logic.

I haven't yet looked at the tests, but when I run `flutter test` I get some failures.

File: app/lib/features/bluetooth/backend/bluetooth_low_energy/ble_manager.dart
Comment (NobodyForNothing) [reply app/lib/features/bluetooth/backend/bluetooth_low_energy/ble_manager.dart:12]: Sorry if I was nitpicky. I really appreciate all contributions and wouldn't consider this some kind of blocker at all. Since you opened so many issues, translated the app to a new language and made such a large PR here I assumed/hoped you wanted to get more involved.

I don't want there to be any pressure, this is a passion project, so I will make this review as quick as possible and get this merged ;). I really appreciate everything you have done so far.

**Comment by NobodyForNothing** [comment]
> Afaik it should be possible to add a ref: $GITHUB_REF or ref: ${{ github.ref }} line to the checkout step?

Thanks for the suggestion. I played around a bit more in another repo and found out I had to run a `gh pr`checkout after cloning. /build should work now in case you want to iterate some more.

**Comment by derdilla** [comment]
> you have to request permissions during runtime like fbp does? 

@pimlie yes, permissions are generally requested at runtime. This could either be done by the bluetooth implementations or in a seperate step using something like [permission_handler](https://pub.dev/packages/permission_handler).

> sorry for the slow updates

Don't stress yourself about this. Since testing on android isn't convenient for you (and you have no hot reload) I can also do this in another PR and save you some time.

**Comment by derdilla** [comment]
@pimlie I looked into the ble lines you send and they look like enough, potentially even too much permissions (iirc modern android doesn't need that many location permissions). I only glanced at the rest of their code, but it looks like all permissions are [requested at once](https://github.com/yanshouwang/bluetooth_low_energy/blob/e24d1bc302a15dbb1232ad0f3db30406f5776060/bluetooth_low_energy_android/android/src/main/kotlin/dev/hebei/bluetooth_low_energy_android/MyCentralManager.kt#L133C28-L133C46), which fits the observed behavior of opening the app-settings page. 

Open an issue if you want, I think that's something the maintainer should at least be aware of.  This is however not blocking for us since we can request the permissions ourself (architecture wise both sides have valid arguments).

In other news: Due to personal reasons I will have less/no time next week, so sorry if you don't get immediate responses.

**Comment by derdilla** [comment]
@pimlie glad to hear my retrying had at least some use. You probably already thought of this, but I just had the idea of "just" ignoring disconnects/retrying until at least one measurement is read.

**Comment by 5FeetUnder** [comment]
> Ideally (but not necessarily):
> 
> * someone with another ble device tests this

I just tested 4bff687 with the Beurer BM 64 and it offered me all measurements that were on the device to select from (it is apparently of the variety "send all stored measurements after taking a measurement").

I have ideas about usability, but those are probably better suited as feature requests after this PR got merged, since I don't want to blow its scope up more than necessary.

**Comment by derdilla** [comment]
@pimlie your description sounds like you are building from a stable flutter, and try to build beta-flutter code.

In that case you can either use beta flutter or use the latest android/ and pubspec. files from main branch.

File: app/test/features/bluetooth/ui/measurement_multiple_test.dart
Comment (derdilla) [inline app/test/features/bluetooth/ui/measurement_multiple_test.dart:13]: <details><summary>I'm still getting a failure:</summary>
<p>

```
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure was thrown running a test:
Expected: <1>
  Actual: <0>

When the exception was thrown, this was the stack:
#4      main.<anonymous closure> (file:///home/derdilla/Coding/Flutter/blood_pressure_app/app/test/features/bluetooth/ui/measurement_multiple_test.dart:83:5)
<asynchronous suspension>
#5      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:189:15)
<asynchronous suspension>
#6      TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:1032:5)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

This was caught by the test expectation on the following line:
  file:///home/derdilla/Coding/Flutter/blood_pressure_app/app/test/features/bluetooth/ui/measurement_multiple_test.dart line 83
The test description was:
  should show everything and be interactive
════════════════════════════════════════════════════════════════════════════════════════════════════
00:22 +131 -3: /home/derdilla/Coding/Flutter/blood_pressure_app/app/test/features/bluetooth/ui/measurement_multiple_test.dart: should show everything and be interactive [E]                                                                     
  Test failed. See exception logs above.
  The test description was: should show everything and be interactive
```

</p>
</details> 

If you don't want to learn the test framework that's fine. Just comment this out and I will do appropriate testing when writing the logic out of the UI.

File: app/test/features/bluetooth/logic/device_scan_cubit_test.dart
Comment (derdilla) [inline app/test/features/bluetooth/logic/device_scan_cubit_test.dart:1]: <details><summary>Test failure: see what I wrote in measurement_multiple test</summary>
<p>

```
000:07 +123 -1: /home/derdilla/Coding/Flutter/blood_pressure_app/app/test/features/bluetooth/logic/device_scan_cubit_test.dart: finds and connects to devices [E]                                                                                 
  Expected: should emit an event that <Instance of 'DeviceListAvailable'>
    Actual: <Instance of '_BroadcastStream<DeviceScanState>'>
     Which: emitted • Instance of 'SingleDeviceAvailable'
              which emitted an event that is not an instance of 'DeviceListAvailable'
  
  package:matcher                                                  expectLater
  package:flutter_test/src/widget_tester.dart 517:25               expectLater
  test/features/bluetooth/logic/device_scan_cubit_test.dart 64:11  main.<fn>
  ===== asynchronous gap ===========================
  dart:async                                                       _CustomZone.registerUnaryCallback
  test/features/bluetooth/logic/device_scan_cubit_test.dart 61:5   main.<fn>
  

To run this test again: /home/derdilla/bin/apps/flutter/bin/cache/dart-sdk/bin/dart test /home/derdilla/Coding/Flutter/blood_pressure_app/app/test/features/bluetooth/logic/device_scan_cubit_test.dart -p vm --plain-name 'finds and connects to devices'
00:07 +123 -2: /home/derdilla/Coding/Flutter/blood_pressure_app/app/test/features/bluetooth/logic/device_scan_cubit_test.dart: recognizes devices [E]                                                                                            
  Expected: should emit an event that <Instance of 'DeviceSelected'>
    Actual: <Instance of '_BroadcastStream<DeviceScanState>'>
     Which: emitted • Instance of 'SingleDeviceAvailable'
              which emitted an event that is not an instance of 'DeviceSelected'
  
  package:matcher                                                   expectLater
  package:flutter_test/src/widget_tester.dart 517:25                expectLater
  test/features/bluetooth/logic/device_scan_cubit_test.dart 113:11  main.<fn>
  ===== asynchronous gap ===========================
  dart:async                                                        _CustomZone.registerUnaryCallback
  test/features/bluetooth/logic/device_scan_cubit_test.dart 107:5   main.<fn>
```

</p>
</details>

**Review by derdilla — CHANGES_REQUESTED** [review]
Just a few things before we can get this merged: Test failures that need to be disabled, and potential bugs you should check again.

I'm fine with keeping the UI stuff, some doc changes, package extraction, and proper testing to other PRs.

**Comment by derdilla** [comment]
> Can we indicate that correctly in the pubspec maybe?

@pimlie, that seems reasonable. The CI should probably also run with the oldest supported version. Sorry for the confusion, I really should get out of the habit of having undocumented things like this. You are the first major contributor so feedback like this is highly appreciated.

As soon as this is ready for merge: Do you prefer your PR rebased (main-branch gets 90 nicely named conventional-commits) or squashed (one clean commit and we rely on gh if we ever need details)?

**Comment by derdilla** [comment]
@pimlie quick question: What's the state of this?

**Review by derdilla — APPROVED** [review]
LGTM👍.

iirc correctly you wanted to have a beta testing phase after landing this. I can definitely make a pre-release on GitHub and setup a new beta-track on the play store. I don't know how betas on f-droid work or how we even get testers on this.

We could of course do the annoying thing and ping the people that already helped with Bluetooth or showed interested in it. It would be quite beneficial if you had a way for me to reach out to potential testers 🙏.

**Pull request merged**
